docs/src/main/asciidoc/security-authorize-web-endpoints-reference.adoc

```diff
@@ -1078,6 +1079,163 @@ public @interface CanWrite {
+
+[source,java]
+----
+package org.acme.security.permission;
+
+import java.security.Permission;
+
+public class BeanParamPermission extends Permission {
+
+    private final String permissionName;
+    private final String customAuthorization;
+    private final String userName;
+    private final String queryParam;
+
+    public BeanParamPermission(String permissionName, String customAuthorizationHeader, String name, String query) {
+        super(permissionName);
+        this.permissionName = permissionName;
+        this.customAuthorization = customAuthorizationHeader;
+        this.userName = name;
+        this.queryParam = query;
+    }
+
+    @Override
+    public boolean implies(Permission possessedPermission) {
+        boolean permissionNameMatches = permissionName.equals(possessedPermission.getName());
+        boolean queryParamAllowedForPermissionName = checkQueryParams(queryParam);
+        boolean usernameWhitelisted = isUserNameWhitelisted(userName);
+        boolean customAuthorizationMatches = checkCustomAuthorization(customAuthorization);
+        return permissionNameMatches && queryParamAllowedForPermissionName && usernameWhitelisted
+                    && customAuthorizationMatches;
+    }
+
+    ...
+}
+----
+
+Permission `BeanParamPermission` is required to access the `SimpleResource#sayHello` endpoint.
+Access will only be granted if you add `possessedPermission` from the previous example to your `SecurityIdentity`:
+
+[source,java]
+----
+package org.acme.security.permission;
+
+import java.security.Permission;
+
+import jakarta.enterprise.context.ApplicationScoped;
+
+import io.quarkus.security.StringPermission;
+import io.quarkus.security.identity.AuthenticationRequestContext;
+import io.quarkus.security.identity.SecurityIdentity;
+import io.quarkus.security.identity.SecurityIdentityAugmentor;
+import io.quarkus.security.runtime.QuarkusSecurityIdentity;
+import io.smallrye.mutiny.Uni;
+
+@ApplicationScoped
+public class PermissionSecurityIdentityAugmentor implements SecurityIdentityAugmentor {
+
+    @Override
+    public Uni<SecurityIdentity> augment(SecurityIdentity securityIdentity,
+            AuthenticationRequestContext authenticationRequestContext) {
+        var possessedPermission = createPossessedPermission(securityIdentity);
+        var augmentedIdentity = QuarkusSecurityIdentity
+                .builder(securityIdentity)
+                .addPermissionChecker(requiredPermission -> Uni
+                        .createFrom()
+                        .item(requiredPermission.implies(possessedPermission)))
+                .build();
+        return Uni.createFrom().item(augmentedIdentity);
+    }
+
+    private static Permission createPossessedPermission(SecurityIdentity securityIdentity) {
+        // replace next line with your business logic
+        return securityIdentity.isAnonymous() ? new StringPermission("ping") : new StringPermission("read");
```

Comment:
This is backwards, we don't care whether a required permission implies a possessed permission. It's meant to be the other way around.

Look, I'm sorry but I think this is a confusing API.

You have something that is a `PermissionChecker`, which is a `Function<Boolean,Permission>`.  The problem is that we picked `Permission` which is not the right thing for the job, and we're confusing the permission with the permission checker.

If Permission had a `boolean isGranted()` method, everything would be logical, and we'd be able to write the much more intuitive following code:

```java
// frankly I'd get rid of the Permission supertype here, I don't think it does anything for us
public abstract class QuarkusPermission extends Permission {
 public abstract boolean isGranted(); // perhaps add SecurityIdentity as parameter?
}

// ...
public class BeanParamPermission extends Permission {
    private final String permissionName;
    private final String customAuthorization;
    private final String userName;
    private final String queryParam;
    public BeanParamPermission(String permissionName, String customAuthorizationHeader, String name, String query) {
         super(permissionName);
        this.permissionName = permissionName;
        this.customAuthorization = customAuthorizationHeader;
        this.userName = name;
        this.queryParam = query;
    }
    @Override
    public boolean isGranted() {
        boolean queryParamAllowedForPermissionName = checkQueryParams(queryParam);
        boolean usernameWhitelisted = isUserNameWhitelisted(userName);
        boolean customAuthorizationMatches = checkCustomAuthorization(customAuthorization);
        return queryParamAllowedForPermissionName && usernameWhitelisted
                    && customAuthorizationMatches;
    }
}
```

Then, you can define a default `PermissionChecker` (in Quarkus, not needed to be written by the user) that does this for every `SecureIdentity`:

```java
        var augmentedIdentity = QuarkusSecurityIdentity
                .builder(securityIdentity)
                .addPermissionChecker(requiredPermission -> Uni
                        .createFrom()
                        .item(requiredPermission.isGranted()))
                .build();
```

Now, if you really insist on keeping `Permission`, for some reason, I would say it's pretty obvious to me that there are:
- global permissions that a user has, perhaps `String` permissions, that do not vary based on the request, and you might add `QuarkusSecurityIdentity.builder(id).permissions(Permission...)` to set them, and you can query them for each request, and use `globalPermission.implies(requestedPermission)` to figure out if you have a global permission, and if not, then
- per-request permissions that depend on request things, like path, current user, parameters, in which case we use `PermissionChecker` and see if that permission is granted for this request, and those can't imply other permissions

I think this API would make a lot more sense to people like me.

Reply:
> This is backwards, we don't care whether a required permission implies a possessed permission. It's meant to be the other way around.

I realize that required permission must be subset of the possessed permission, but considering you are writing here both sides (possessed and required) then you are the own who decides whether `possessedPermission equals requiredPermission` is same as `requiredPermission implies possessedPermission` because equality can mean two-way implication (r -> p & p -> r) depending on axiomatic system you use. Anyway, I'll remove it and I suggest to only discuss about remaining of your comments.

> You have something that is a PermissionChecker, which is a Function<Boolean,Permission>.

BTW typo, its `Function<Permission, Uni<Boolean>>`, but you know that

> If Permission had a boolean isGranted() method, everything would be logical, and we'd be able to write the much more intuitive following code:

Yes, I agree that `QuarkusPermission` would simplify that. I don't know how else rewrite docs or address comments so I'll implement it.

Reply:
Hi @FroMage Can you copy it to your issue please, I'm planning to comment there as well with another suggestion

Reply:
@sberyozkin I'd also appreciate to resolve this in a separate issue to keep this small, but I think I need somehow deal with @FroMage docs request. Honestly I am out of ideas, so I don't know how to rewrite it without https://github.com/quarkusio/quarkus/issues/43238. I suppose I could at least dropped augmentor and return `that` to this example to address that comment? Don't know.

Reply:
@michalvavrik Hi Michal, do you plan to do it in another PR, is what you are thinking about ?

Reply:
> Now, if you really insist on keeping Permission, for some reason, I would say it's pretty obvious to me that

that was discussed at length here https://groups.google.com/g/quarkus-dev/c/ADY6k-Kwv-Y/ and Stuart explained his ideas here, so I think it is possible to implement `QuarkusPermission` on top of the `java.security.Permission`.

Reply:
>  @michalvavrik Hi Michal, do you plan to do it in another PR, is what you are thinking about ?

yes, right after this PR, but I don't know how to deal with docs ATM, see above comments and previous threads....

Reply:
I'll try yet another docs version to settle this, there used to be button in Google search "I feel lucky".... :-)

Reply:
@michalvavrik Sure, better iterate with the `QuarkusPermission` simplification in the follow up PR

Reply:
Hello @FroMage ,

thanks for the patience, I think I have reasonable example now, but feel free to fight back, I think all your comments are spot on.  Please have a look.

@sberyozkin maybe re-check if changes in that one example are OK with you, thanks as well.